source/s3_meta_request.c

```diff
@@ -842,20 +832,269 @@ static void s_s3_meta_request_send_request_finish_default(
     }
 }
 
-/* Push a request description into the retry queue.  This assumes ownership of the request desc. */
-void aws_s3_meta_request_queue_retry(struct aws_s3_meta_request *meta_request, struct aws_s3_request *request) {
+static void s_s3_meta_request_retry_ready(struct aws_retry_token *token, int error_code, void *user_data) {
+    AWS_PRECONDITION(token);
+    (void)token;
+
+    struct aws_s3_request *request = user_data;
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
     AWS_PRECONDITION(meta_request);
-    AWS_PRECONDITION(request && request->meta_request == meta_request);
+
+    /* If we couldn't retry this request, then bail on the entire meta request. */
+    if (error_code != AWS_ERROR_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not retry request %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            error_code,
+            aws_error_str(error_code));
+
+        aws_s3_meta_request_finish(meta_request, NULL, 0, error_code);
+        goto clean_up;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p Meta request pushing request %p with token %p into retry queue",
+        (void *)meta_request,
+        (void *)request,
+        (void *)request->retry_token);
+
+    /* Push the request into the retry queue so that it can actually be retried. */
+    aws_s3_meta_request_retry_queue_push(meta_request, request);
+
+    /* Tell the client we have additional work that can be done. */
+    aws_s3_meta_request_schedule_work(meta_request);
+
+clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_release(meta_request);
+}
+
+static void s_s3_meta_request_queue_retry_with_token(
+    struct aws_retry_strategy *retry_strategy,
+    int error_code,
+    struct aws_retry_token *token,
+    void *user_data) {
+
+    AWS_PRECONDITION(retry_strategy);
+    (void)retry_strategy;
+
+    struct aws_s3_request *request = user_data;
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    AWS_PRECONDITION(meta_request);
+
+    if (error_code != AWS_ERROR_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not get retry token request %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_ref_clean_up;
+    }
+
+    AWS_ASSERT(token);
+
+    enum aws_retry_error_type error_type = AWS_RETRY_ERROR_TYPE_TRANSIENT;
+
+    if (error_code == AWS_ERROR_S3_NO_PART_BUFFER) {
```

Comment:
switch maybe

Reply:
err I'm confused, doesn't the previous if make this entire if-chain useless?

Reply:
Yes, yes it it does.  It's meant to be looking at the error code for the request here, not the error code passed in.  Thanks for catching that!